File: backend/scripts/test_auth.mjs
const BASE = 'http://127.0.0.1:8000';
const TEST = { username: 'testuser99', email: 'testuser99@example.com', password: 'testpass123' };

async function post(path, body) {
  const res = await fetch(`${BASE}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  const data = await res.json().catch(() => ({}));
  return { status: res.status, data };
}

async function main() {
  console.log('Health check...');
  try {
    await fetch(`${BASE}/health`);
  } catch {
    console.error('FAIL: Backend not running at', BASE);
    process.exit(1);
  }

  console.log('Signup...');
  let r = await post('/api/auth/signup', TEST);
  if (r.status === 201) {
    console.log('OK signup', r.data.user?.username);
  } else if (r.status === 400 && String(r.data.detail).includes('already')) {
    console.log('User exists, continuing...');
  } else {
    console.error('FAIL signup', r.status, r.data);
    process.exit(1);
  }

  console.log('Login...');
  r = await post('/api/auth/login/json', {
    username_or_email: TEST.username,
    password: TEST.password,
  });
  if (r.status !== 200) {
    console.error('FAIL login', r.status, r.data);
    process.exit(1);
  }
  console.log('OK login', r.data.user?.email);
  console.log('All tests passed.');
}

main();

File: backend/scripts/test_auth.py
"""Test signup and login against running API."""
import json
import urllib.error
import urllib.request

BASE = "http://127.0.0.1:8000"
TEST_USER = {
    "username": "testuser99",
    "email": "testuser99@example.com",
    "password": "testpass123",
}


def post(path, data):
    req = urllib.request.Request(
        f"{BASE}{path}",
        data=json.dumps(data).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as res:
            return res.status, json.loads(res.read().decode())
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        try:
            detail = json.loads(body).get("detail", body)
        except Exception:
            detail = body
        return e.code, detail


def get(path, token):
    req = urllib.request.Request(
        f"{BASE}{path}",
        headers={"Authorization": f"Bearer {token}"},
    )
    with urllib.request.urlopen(req, timeout=10) as res:
        return json.loads(res.read().decode())


if __name__ == "__main__":
    print("1. Signup...")
    code, result = post("/api/auth/signup", TEST_USER)
    print(f"   Status: {code}")
    if code != 201:
        print(f"   FAIL: {result}")
        if code == 400 and "already" in str(result).lower():
            print("   (User exists — trying login instead)")
        else:
            raise SystemExit(1)
    else:
        token = result["access_token"]
        print(f"   OK user id={result['user']['id']} username={result['user']['username']}")

    print("2. Login...")
    code, result = post(
        "/api/auth/login/json",
        {"username_or_email": TEST_USER["username"], "password": TEST_USER["password"]},
    )
    print(f"   Status: {code}")
    if code != 200:
        print(f"   FAIL: {result}")
        raise SystemExit(1)
    token = result["access_token"]
    print(f"   OK token received")

    print("3. Me...")
    me = get("/api/auth/me", token)
    print(f"   OK email={me['email']} role={me['role']}")
    print("\nAll auth tests passed.")
